### PR TITLE
Adds int64 support for SecureNN

### PR DESCRIPTION
**Pull request by mortendahl**
closes https://github.com/mortendahl/tf-encrypted/issues/1
closes https://github.com/mortendahl/tf-encrypted/issues/202

### CODE REVIEW

File: tf_encrypted/protocol/pond.py [hunk 1]
```diff
@@ -3512,3 +3510,22 @@ def _cast_backing_public(
             y_on_1 = x_on_1.cast(backing_dtype)
 
         return PondPublicTensor(prot, y_on_0, y_on_1, x.is_scaled)
+
+
+def _cast_backing_private(prot: Pond, x: PondPrivateTensor, backing_dtype) -> PondPrivateTensor:
```
Comment (justin1121) [inline tf_encrypted/protocol/pond.py]: Wonder if this should be moved to securenn if its not being used elsewhere.

File: tf_encrypted/protocol/securenn/securenn.py [hunk 2]
```diff
@@ -458,50 +443,67 @@ def _bits_public(prot, x: PondPublicTensor,
...
 
     with tf.name_scope('lsb'):
 
         with tf.name_scope('lsb_mask'):
 
             with tf.device(prot.server_2.device_name):
-                x_raw = y.backing_dtype.sample_uniform(y.shape)
-                xbits_raw = x_raw.to_bits(factory=prot.prime_factory)
-                xlsb_raw = xbits_raw[..., 0]
+                r0_raw = odd_dtype.sample_uniform(x.shape)
```
Comment (justin1121) [inline tf_encrypted/protocol/securenn/securenn.py]: Curious why not just `_share`?

File: tf_encrypted/tensor/int100.py [hunk 3]
```diff
@@ -379,6 +380,10 @@ def truncate(self, amount, base=2):
     def right_shift(self, bitlength):
         return self.truncate(bitlength, 2)
 
+    def cast(self, factory):
```
Comment (justin1121) [inline tf_encrypted/tensor/int100.py]: This is saying int100tensors don't need to be casted so just return ourself?

Comment (mortendahl) [reply tf_encrypted/tensor/int100.py]: the assertion makes sure that you can only cast int100 it to int100; it's a bit ugly and should be cleanup up once we have more advanced types for tensors!

Comment (mortendahl) [reply tf_encrypted/tensor/int100.py]: for now it's needed to use the same code for securenn on int64 and int100 (where the former is cast to the odd tensor)

File: tf_encrypted/protocol/securenn/securenn.py (hunk 2 above)
Comment (mortendahl) [reply tf_encrypted/protocol/securenn/securenn.py]: good catch, believe we could change it now if that's better.

motivation was that `_share` works differently: it draws one sample and subtracts that from the value. here we are essentially sharing a random value, so doing it this way gives exactly the same; however, no subtraction is involved. in the end I implemented subtraction for odd tensors as well so `_share` could be used, yet this is likely how we'll share random tensors after introducing seeding.

File: tf_encrypted/protocol/pond.py (hunk 1 above)
Comment (mortendahl) [reply tf_encrypted/protocol/pond.py]: yeah I like that, not least since it's half-baked for now. will move both public and private casting

Comment (mortendahl) [reply tf_encrypted/protocol/pond.py]: done

**Pull request merged**